subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheRemoteBuildScriptIntegrationTest.groovy

```diff
@@ -0,0 +1,113 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+import org.gradle.test.fixtures.server.http.HttpServer
+import org.junit.Rule
+import spock.lang.Issue
+
+class ConfigurationCacheRemoteBuildScriptIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    @Rule
+    HttpServer server = new HttpServer()
+
+    def scriptUrl
+    def scriptFile
+    def configurationCache
+    def scriptName
+
+    def setup() {
+        server.start()
+
+        scriptName = "remote-script.gradle"
+        scriptUrl = "${server.uri}/${scriptName}"
+        scriptFile = file("remote-script.gradle") << """
+            println 'loaded remote script'
+        """
+        server.expectGet("/$scriptName", scriptFile)
+
+        buildFile << """
+            apply from: '$scriptUrl'
+            task ok
+        """
+
+        configurationCache = newConfigurationCacheFixture()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/23273")
+    def "invalidates cache if remote script was changed"() {
+        when:
+        configurationCacheRun 'ok'
+
+        then:
+        configurationCache.assertStateStored()
+        scriptFile << """
+            print 'update remote script'
+        """
+
+        and:
+        server.expectGet("/$scriptName", scriptFile)
+
+        //below the call from DefaultCacheAwareExternalResourceAccessor
+        server.expectHead("/$scriptName", scriptFile)
+        server.expectGet("/$scriptName", scriptFile)
+        configurationCacheRun 'ok'
+
+        then:
+        output.contains("Calculating task graph as configuration cache cannot be reused because remote script $scriptUrl has changed.")
+        configurationCache.assertStateStored()
+    }
+
+    def "reuse cache if remote script is up to date"() {
+        when:
+        configurationCacheRun 'ok'
+
+        then:
+        configurationCache.assertStateStored()
+
+        and:
+        server.expectGet("/$scriptName", scriptFile)
+        configurationCacheRun 'ok'
+
+        then:
+        output.contains("Configuration cache entry reused.")
+    }
+
+    def "reuse cache for offline build"() {
+        when:
+        configurationCacheRun 'ok'
+
+        then:
+        configurationCache.assertStateStored()
+
+        and:
+        executer.withArgument("--offline")
+        configurationCacheRun 'ok'
+
+        and:
+        scriptFile << """
+            print 'update remote script'
+        """
+
+        and:
+        executer.withArgument("--offline")
+        configurationCacheRun 'ok'
+
+        then:
+        output.contains("Configuration cache entry reused.")
```

Comment:
🧪 Preferably, this type of assertion would be expressed via the `configurationCache` fixture, in this case `configurationCache.assertStateLoaded()`.

Reply:
Done